Why "createReducer" return type isn't a deox/dist/types/Reducer?

For those who received a incomplete issue in email, I am sorry, I have hit enter accidentally before finishing.

Now, to the issue:

I am trying to pass a Reducer as a parameter to a method, but I get a compile error about the types.

Please let me know if I am using the library wrong:

> Error:(61, 27) TS2345: Argument of type '(state: DeepImmutableArray<Param> | undefined, action: { type: string; payload: Param; } | { type: string; payload: { index: number; param: Param; }; } | { type: string; payload: { fromIndex: number; toIndex: number; }; }) => Param[] | DeepImmutableArray<...>' is not assignable to parameter of type 'Reducer<DeepImmutableArray<Param>, { type: string; payload: Param; } | { type: string; payload: { index: number; param: Param; }; } | { type: string; payload: { fromIndex: number; toIndex: number; }; }>'.
>   Types of parameters 'state' and 'prevState' are incompatible.
>     Property '[Symbol.iterator]' is missing in type 'DeepImmutableObject<DeepImmutableArray<Param>>' but required in type 'DeepImmutableArray<Param>'.

```typescript
export function createParamTableReducer(
  actions: ParamTableActionCreators
): Reducer<Param[]> {
  const deoxReducer = createReducer([] as Param[], (handle) => [
    handle(actions.addParam, (state, { payload }) =>
      produce(state, (draft: Draft<Param[]>) => {
        draft.push(payload);
      })
    ),
    handle(actions.removeParam, (state, { payload }) =>
      produce(state, (draft: Draft<Param[]>) => {
        draft.splice(payload.index, 1);
      })
    ),
    handle(actions.updateParam, (state, { payload }) =>
      produce(state, (draft: Draft<Param[]>) => {
        draft[payload.index] = payload.param;
      })
    ),
    handle(actions.reorderParams, (state, { payload }) =>
      produce(state, (draft: Draft<Param[]>) => {
        draft.splice(payload.fromIndex, 1);
        draft.splice(payload.toIndex, 0, state[payload.fromIndex]);
      })
    ),
  ]);

  return normalizeReducer(deoxReducer);
}
```

```typescript
import { Reducer } from "deox/dist/types";
import { Action, Reducer as ReduxReducer } from "redux";

export function normalizeReducer<T, A extends Action>(
  deoxReducer: Reducer<T, A>
): ReduxReducer<T, A>
```

Why doesn't "createReducer" uses the existing Reducer type as a returnType?

@thebrodmann I am having a problem while pushing the changes.

```
Test Suites: 6 passed, 6 total
Tests:       202 passed, 202 total
Snapshots:   6 obsolete, 6 written, 95 passed, 101 total
Time:        27.952s
Ran all test suites.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deox@1.4.0 test-dts: `jest -c dts-jest.config.js "--bail"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deox@1.4.0 test-dts script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Marcio\AppData\Roaming\npm-cache\_logs\2019-06-04T19_47_36_896Z-debug.log
husky > pre-push hook failed (add --no-verify to bypass)
error: failed to push some refs to 'https://github.com/LouizFC/deox.git'

```

--- a/src/types.ts
+++ b/src/types.ts
@@ -46,7 +46,7 @@ export type Reducer<TState, TAction> = (
 ) => DeepImmutable<TState> | TState
 
 export type ActionType<
-  T extends ActionCreator<AnyAction> | Handler<any, Action<any>>
+  T extends ActionCreator<AnyAction> | Reducer<any, Action<any>>
 > = T extends ActionCreator<AnyAction>
   ? ReturnType<T>
-  : (T extends Handler<any, infer U> ? U : never)
+  : (T extends Reducer<any, infer U> ? U : never)

